### PR TITLE
core: make ModulePass frozen

### PR DESCRIPTION
**Pull request by superlopuh**
Mainly through using n-tuples instead of lists.

This makes it much easier to reason about pass pipelines in general, but also for the interactive app specifically. If we can assume that all passes are immutable, we don't have to reconstruct them every time just to be safe.

### CODE REVIEW

**Review by compor — COMMENTED** [review]
What do you mean by reason about the pass pipeline?

Did we dynamically change options and/or the behaviour of the passes so far?
Did we introspect it in any way?

**Comment by superlopuh** [comment]
We never change options or behaviour, so this is basically an NFC. My proposal is to encode it as part of the type, so that we don't do it in the future. Today, they are difficult to introspect for this reason, and not safe to reuse, leading to a bit of weirdness on the interactive side where we initialise a new pass every time. When trying to exhaustively search the space of possible IRs given a set of rewrite passes, initing a new one every time adds up quickly.

**Comment by compor** [comment]
The introspection aspect is not clear at all to me and how this would help this (maybe slightly), but I do get the interactive use case part. Thanks for elaborating.

**Review by compor — APPROVED** [review]

**Review by PapyChacal — APPROVED** [review]

**Review by dshaaban01 — APPROVED** [review]

File: xdsl/transforms/arith_add_fastmath.py [hunk 1]
```diff
@@ -32,18 +32,20 @@ class AddArithFastMathFlags(RewritePattern):
...
-        | arith.Minimumf
-        | arith.Maximumf,
+        op: (
+            arith.Addf
+            | arith.Subf
+            | arith.Mulf
+            | arith.Divf
+            | arith.Minimumf
+            | arith.Maximumf
+        ),
```
Comment (math-fehr) [inline xdsl/transforms/arith_add_fastmath.py:42]: I think that is an unrelated change? At least, that's not usually the style we are using

**Review by math-fehr — APPROVED** [review]
That makes sense to me!
I'm just not a fan of the new syntax for larger Union types!

**Review by AntonLydike — APPROVED** [review]
LGTM

File: xdsl/transforms/arith_add_fastmath.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/transforms/arith_add_fastmath.py:42]: There's some weirdness in Black formatting, I didn't make this change manually. I've seen it in some other PRs, I think I have a situation where the formatter in VSCode and in precommit are slightly different, and both slightly ambiguous

Comment (superlopuh) [reply xdsl/transforms/arith_add_fastmath.py:42]: #2271
